src/main/java/org/victorrobotics/frc/dtlib/drivetrain/swerve/DTSwerveDrive.java

```diff
@@ -0,0 +1,154 @@
+
+    private Field2d       virtualField;
+    private ChassisSpeeds previousSpeeds;
+    private ChassisSpeeds currentSpeeds;
+
+    protected DTSwerveDrive(DTSwerveModule... modules) {
+        Objects.requireNonNull(modules);
+        if (modules.length < 2) {
+            throw new IllegalArgumentException("Swerve drive requires at least 2 wheels");
+        }
+        for (DTSwerveModule m : modules) {
+            Objects.requireNonNull(m);
+        }
+        this.modules = modules.clone();
+
+        setSubsystem("Swerve Drive");
+        setName(getSubsystem());
+
+        initializeHardware();
+
+        Translation2d[] wheelLocations = new Translation2d[modules.length];
+        positions = new SwerveModulePosition[modules.length];
+        for (int i = 0; i < wheelLocations.length; i++) {
+            wheelLocations[i] = modules[i].getLocation();
+            positions[i] = modules[i].getPosition();
+        }
+
+        kinematics = new SwerveDriveKinematics(wheelLocations);
+        poseEstimator = new SwerveDrivePoseEstimator(kinematics, getGyroAngle(), positions,
+                new Pose2d());
+
+        centerOfRotation = new Translation2d();
+        accelerationLimit = new DTAccelerationLimit();
+        currentSpeeds = new ChassisSpeeds();
+    }
+
+    public void initializeHardware() {
+        for (DTSwerveModule module : modules) {
+            module.initializeHardware();
+        }
+    }
+
+    public final void setCenterOfRotation(Translation2d newCenterOfRotation) {
+        centerOfRotation = newCenterOfRotation;
+    }
+
+    public void setAccelerationLimit(DTAccelerationLimit limit) {
+        accelerationLimit = limit;
+    }
+
+    public void setVelocityLimit(DTVelocityLimit limit) {
+        velocityLimit = limit;
+    }
+
+    public void setFieldRelative() {
+        isFieldRelative = true;
+    }
+
+    public void setRobotRelative() {
+        isFieldRelative = false;
+    }
+
+    protected abstract Rotation2d getGyroAngle();
+
+    public final void driveVelocity(double vx, double vy, double vr) {
+        driveVelocity(vx, vy, vr, centerOfRotation);
+    }
+
+    public void driveVelocity(double vx, double vy, double vr, Translation2d centerOfRotation) {
+        previousSpeeds = currentSpeeds;
+
+        if (isFieldRelative) {
+            currentSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(vx, vy, vr, getGyroAngle());
+        } else {
+            currentSpeeds = new ChassisSpeeds(vx, vy, vr);
+        }
+
+        velocityLimit.apply(currentSpeeds);
+        accelerationLimit.apply(currentSpeeds, previousSpeeds);
+
+        SwerveModuleState[] newStates = kinematics.toSwerveModuleStates(currentSpeeds,
+                centerOfRotation);
+        setStates(newStates);
+    }
+
+    public final void setStates(SwerveModuleState... states) {
+        if (states.length != modules.length) {
+            throw new DTIllegalArgumentException("received " + states.length + " module states for "
```

Comment:
We will need to be sure that robots catch exceptions

Reply:
We will have to define behavior when an exception is thrown. In case of a Command, for example, it might simply be canceled immediately and not restarted. For a subsystem, maybe it goes through its initialization again?